### DISCUSSION TITLE
How to using SQL get the previous month aggregate value such as sum?

### DISCUSSION THREAD

**Discussion by hengaini2055** [ClickHouse/ClickHouse#69129] [Q&A] [answered]
[Link to stackoverflow](https://stackoverflow.com/questions/78936766/how-to-define-a-metric-using-sql-to-get-last-month-sum-value-in-clickhouse)

**Comment by den-crane** [comment]
Last or previous?

```sql

 SELECT
     pt,
     objectCode,
     MonthValue,
     lagInFrame(MonthValue) OVER (PARTITION BY objectCode ORDER BY pt ASC ROWS
                  BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) AS prevMonthValue,
     first_value(MonthValue) OVER (PARTITION BY objectCode ORDER BY pt desc) AS lastMonthValue,                  
 FROM 
   (select toStartOfMonth(Period) as pt, objectCode, sum(amount) as MonthValue
   from posts group by pt, objectCode)
 ORDER BY
     pt,
     objectCode;
```

running/accumulated ?

```sql
 SELECT
     pt,
     objectCode,
     MonthValue,
     sum(MonthValue) OVER (PARTITION BY objectCode ORDER BY pt ) runningSum
 FROM 
   (select toStartOfMonth(Period) as pt, objectCode, sum(amount) as MonthValue
   from posts group by pt, objectCode)
 ORDER BY
     pt,
     objectCode;
```

**Comment by hengaini2055** [comment]
When the inner sql subexpress gets all Months, the SQL is right:
![image](https://github.com/user-attachments/assets/701ae751-ae72-459e-8775-8659335d0ff9)
![image](https://github.com/user-attachments/assets/c186ea1c-cb95-4dd6-9c1a-74fabe168e66)
When the data for a objectCode for example 'D000000147' have only two month data, the SQL get the prevMonth is the prev row value, Not prev month value:
![image](https://github.com/user-attachments/assets/b1e57528-a3a2-4cc0-abc8-876a4a322995)
![image](https://github.com/user-attachments/assets/49910df5-f2f9-40bf-baa8-a08ca8ace729)
To my project, to fill all Period for all objectCode is not acceptable, so prev row is not exact prev month?
[clickhouse_posts.csv](https://github.com/user-attachments/files/16830630/clickhouse_posts.csv)

**Reply by den-crane** [reply]
you need to use range frame https://clickhouse.com/docs/en/sql-reference/window-functions#moving--sliding-average-per-10-seconds

```
     any(MonthValue) OVER (PARTITION BY objectCode ORDER BY toRelativeMonthNum(pt) 
          Range BETWEEN 1 PRECEDING AND 1 PRECEDING) AS prevMonthValue
```

https://fiddle.clickhouse.com/87441738-ecf0-42e4-aef7-e8c3b27b24e2

https://fiddle.clickhouse.com/34523546-c5e4-40d0-a922-3ce83df516da

**Reply by den-crane** [reply, marked as answer]
Please don't use pictures. I don't like them.

**Comment by hengaini2055** [comment]
The key is how to replace filter for column 'Period' like DAX funciton `calculate`. When we get thisMonthValue from group by Period, we cann't replace filter for column 'Period' such '2023-03-01' with '2023-02-01', and remain others filters such as objectCode.

**Comment by hengaini2055** [comment]
The function ifs such as `sumif` in a `group by` statement, can only refrerence the other columns and not in `group by` columns.
